File: Observer Pattern/observer.ts
interface Subject {
  registerObserver(O: Observer): void;
  removeObserver(O: Observer): void;
  notifyObservers(Message: String): void;
}

interface Network extends Subject {
  setConnectionType(ConnectionType: String): void;
}

interface Observer {
  update(Message: String);
}

function Network(): Network {
  let observers = <Array<Observer>>[];

  return {
    setConnectionType(type) {
      this.notifyObservers(`The network is now connected via: ${type}`);
    },
    registerObserver(obs) {
      // do not allow an observer to register twice
      if (observers.find(obsEl => obs === obsEl)) return;

      observers = [...observers, obs];
    },
    removeObserver(obs) {
      observers = observers.filter(obsEl => obsEl !== obs);
    },
    notifyObservers(message) {
      observers.forEach(obs => obs.update(message));
    }
  };
}

function Client(nickname: String): Observer {
  return {
    update(message) {
      console.log(`${nickname}: ${message}`);
    }
  };
}

const network = Network();

const client1 = Client("Client 1");
const client2 = Client("Client 2");
const client3 = Client("Client 3");

network.registerObserver(client1);
network.registerObserver(client1);
network.registerObserver(client2);

network.setConnectionType("WiFi");

network.removeObserver(client2);
network.registerObserver(client3);

network.setConnectionType("Ethernet");

File: Observer Pattern/observer.rb
class Subject
  def initialize
    @observers = []
  end

  def register_observer(observer)
    return if @observers.include?(observer)

    @observers << observer
  end

  def remove_observer(observer)
    @observers.reject! { |obs| obs == observer }
  end

  def notify_observers(message)
    @observers.each { _1.update(message) }
  end
end

class Network < Subject
  def set_connection_type(type)
    notify_observers("The network is now connected via: #{type}")
  end
end

class Observer
  def update(_data)
    raise NotImplementedError, "#update must be defined in observer subclasses"
  end
end

class Client < Observer
  def initialize(nickname)
    @nickname = nickname
  end

  def update(message)
    puts "#{@nickname}: #{message}"
  end
end

network = Network.new

client1 = Client.new("Client 1")
client2 = Client.new("Client 2")
client3 = Client.new("Client 3")

network.register_observer(client1)
network.register_observer(client1)
network.register_observer(client2)

network.set_connection_type("WiFi")

network.remove_observer(client2)
network.register_observer(client3)

network.set_connection_type("Ethernet")
